File: backend/routes/admin/heroShots.ts
import { Router, Request, Response } from 'express';
import db from '../../config/database';
import { deleteFromS3 } from '../../config/s3';

const router = Router();

// GET /api/admin/hero-shots — admin view of the 3 slots
router.get('/', async (_req: Request, res: Response) => {
  try {
    const result = await db.query(
      `SELECT slot, image_url, caption, updated_at
         FROM hero_shots
        ORDER BY slot ASC`,
    );
    return res.json({ success: true, hero_shots: result.rows });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// PUT /api/admin/hero-shots/:slot
// Body: { image_url?: string|null, caption?: string|null }
// Updates one slot in place. If image_url changes, the previous S3 file
// is removed (best-effort).
router.put('/:slot', async (req: Request, res: Response) => {
  const slot = Number(req.params.slot);
  if (!Number.isInteger(slot) || slot < 0 || slot > 2) {
    return res.status(400).json({ success: false, error: 'slot must be 0, 1, or 2' });
  }

  const { image_url, caption } = req.body as {
    image_url?: string | null;
    caption?:   string | null;
  };

  const allowed: Record<string, unknown> = {};
  if (image_url !== undefined) allowed['image_url'] = image_url || null;
  if (caption   !== undefined) allowed['caption']   = caption?.trim() ? caption.trim() : null;

  if (Object.keys(allowed).length === 0) {
    return res.status(400).json({ success: false, error: 'No fields to update' });
  }

  try {
    let previousImage: string | null = null;
    if (image_url !== undefined) {
      const before = await db.query('SELECT image_url FROM hero_shots WHERE slot = $1', [slot]);
      previousImage = before.rows[0]?.image_url ?? null;
    }

    const setClauses = Object.keys(allowed).map((k, i) => `${k} = $${i + 1}`);
    const values     = [...Object.values(allowed), slot];
    const result = await db.query(
      `UPDATE hero_shots SET ${setClauses.join(', ')}
        WHERE slot = $${values.length}
       RETURNING slot, image_url, caption, updated_at`,
      values,
    );
    if (!result.rows.length) {
      return res.status(404).json({ success: false, error: 'Slot not found' });
    }

    if (previousImage && previousImage !== (image_url || null)) {
      try { await deleteFromS3(previousImage); }
      catch (s3err) { console.warn('[hero_shots] S3 cleanup failed:', (s3err as Error).message); }
    }

    return res.json({ success: true, hero_shot: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

export default router;

File: backend-golang/internal/handlers/admin/heroshots.go
package admin

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"inanhxink/backend-golang/internal/config"
	"inanhxink/backend-golang/internal/handlers"
)

// GET /api/admin/hero-shots
func ListHeroShots(w http.ResponseWriter, r *http.Request) {
	rows, err := config.DB.Query(context.Background(),
		"SELECT slot, image_url, caption, updated_at FROM hero_shots ORDER BY slot ASC")
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	shots, err := handlers.CollectRows(rows)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true, "hero_shots": shots})
}

// PUT /api/admin/hero-shots/:slot
func UpdateHeroShot(w http.ResponseWriter, r *http.Request) {
	slot, err := strconv.Atoi(chi.URLParam(r, "slot"))
	if err != nil || slot < 0 || slot > 2 {
		handlers.BadRequest(w, "slot must be 0, 1, or 2")
		return
	}

	var fields map[string]any
	if err := json.NewDecoder(r.Body).Decode(&fields); err != nil {
		handlers.BadRequest(w, "Invalid JSON")
		return
	}

	allowed := map[string]bool{"image_url": true, "caption": true}
	setClauses := []string{}
	values := []any{}
	i := 1
	for k, v := range fields {
		if !allowed[k] {
			continue
		}
		if k == "image_url" {
			if s, ok := v.(string); ok && s == "" {
				v = nil
			}
		}
		if k == "caption" {
			if s, ok := v.(string); ok {
				t := strings.TrimSpace(s)
				if t == "" {
					v = nil
				} else {
					v = t
				}
			}
		}
		setClauses = append(setClauses, fmt.Sprintf("%s = $%d", k, i))
		values = append(values, v)
		i++
	}
	if len(setClauses) == 0 {
		handlers.BadRequest(w, "No fields to update")
		return
	}

	var previousImage string
	if _, replacing := fields["image_url"]; replacing {
		config.DB.QueryRow(context.Background(), //nolint
			"SELECT image_url FROM hero_shots WHERE slot = $1", slot).Scan(&previousImage)
	}

	values = append(values, slot)
	rows, err := config.DB.Query(context.Background(),
		fmt.Sprintf("UPDATE hero_shots SET %s WHERE slot = $%d RETURNING slot, image_url, caption, updated_at",
			joinClauses(setClauses), len(values)),
		values...)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, err := handlers.CollectOne(rows)
	if err != nil || row == nil {
		handlers.NotFound(w)
		return
	}

	if newImg, _ := fields["image_url"].(string); previousImage != "" && previousImage != newImg {
		config.DeleteFromS3(previousImage) //nolint
	}

	handlers.OK(w, map[string]any{"success": true, "hero_shot": row})
}
